### PR TITLE
Added deprecation warnings and changed format of the warnings

### PR DESCRIPTION
**Pull request by stinebuu**
As we in NEST 3.0 will replace `allow_offgrid_spikes` with `allow_offgrid_times` in the `spike_generator`, #971, I have added `allow_offgrid_times` to `spike_generator` and added a deprecation warning for `allow_offgrid_spikes`  which is shown the first time a user change the value for `allow_offgrid_spikes`. I added the `allow_offgrid_times` flag so that users can change to this value now if they want to.

I have also changed the way deprecation warnings are displayed on Python level. When a deprecation warning is shown in Python, the deprecation warning is displayed, but also
```
warnings.warn('\n' + text)    # add LF so text starts on new line
```
([this line](https://github.com/nest/nest-simulator/blob/master/pynest/nest/lib/hl_api_helper.py#L91)). So I have added a `format_Warning` function so we just see the actual warning. In addition, deprecation warnings on model types will now only be shown the first time they are created.

Lastly, I have added a deprecation warning to `PrintNetwork` as this will be replaced by `PrintNodes` in NEST 3.0, #679. To let users change to `PrintNodes` if they wish, I have added this function, which right now just call `PrintNetwork()`, but is implemented in the NEST 3 branch.

### CODE REVIEW

File: pynest/nest/lib/hl_api_helper.py [hunk 1]
```diff
@@ -41,8 +41,17 @@
 sps = spp = sr = pcd = kernel = None
 
 # These flags are used to print deprecation warnings only once. The
-# corresponding functions will be removed in the 2.6 release of NEST.
-_deprecation_warning = {'BackwardCompatibilityConnect': True}
+# corresponding functions will be removed in a future release of NEST.
```
Comment (jougs) [inline pynest/nest/lib/hl_api_helper.py]: `<3` removing the version number from this comment ;-)

File: models/spike_generator.h [hunk 2]
```diff
@@ -249,18 +250,22 @@ class spike_generator : public DeviceNode
     bool precise_times_;
 
     //! Allow and round up spikes not on steps; irrelevant if precise_times_
+    bool allow_offgrid_times_;
+    //! Same as allow_offgrid_times_, here for backward compatibility
```
Comment (heplesser) [inline models/spike_generator.h]: Keep only `allow_offgrid_times_` internally. We need backward compatibility of the *user interface*, not of the implementation.

File: models/spike_generator.h [hunk 3]
```diff
@@ -249,18 +250,22 @@ class spike_generator : public DeviceNode
...
     bool allow_offgrid_spikes_;
 
     //! Shift spike times at present to next step
     bool shift_now_spikes_;
 
+    bool deprecation_warning_issued_;
```
Comment (heplesser) [inline models/spike_generator.h]: We may have deprecation warnings also on other parameters in other models in the future. We should find a general way of handling them, so we don't need to implement it in each node class that has a deprecated parameter. Maybe a general class for handling deprecations (essentially a map from names to flags), and then a member of this class in the `Node` base class.

File: models/spike_generator.cpp [hunk 4]
```diff
@@ -178,16 +183,43 @@ nest::spike_generator::Parameters_::set( const DictionaryDatum& d,
...
   const Time& now )
 {
-  const bool flags_changed =
+  const bool allow_offgrid_spikes_changed = updateValue< bool >(
+    d, names::allow_offgrid_spikes, allow_offgrid_spikes_ );
+  const bool allow_offgrid_times_changed =
+    updateValue< bool >( d, names::allow_offgrid_times, allow_offgrid_times_ );
+
+  if ( allow_offgrid_spikes_changed and allow_offgrid_times_changed
+    and not allow_offgrid_spikes_ == allow_offgrid_times_ )
```
Comment (heplesser) [inline models/spike_generator.cpp]: We should simply forbid using old an new parameter name in the same dictionary.

**Comment by jougs** [comment]
@heplesser: I think we should not make the mechanisms for handling deprecations much more sophisticated than they are. In my view most people are anyway not reading them and only change things once they break. Insofar, my take would even be to relax our stability policy for NEST3 and allow breaking changes between releases and only keep things backward compatible in special cases.

**Comment by stinebuu** [comment]
@jougs @heplesser thanks for the reviews! I have now removed most of the references to `allow_offgrid_spikes` internally.

When it comes to the deprecation warnings, I have started working on it, but I am dealing with some linking problems. I haven't been able to work on it that much lately, because of all the work with the benchmarks, so I don't know when I'll be able to work on it again. I'll let you discuss if it is needed. When deprecating a parameter like I have done here, you only need to add a `deprecation_warning_issued_` parameter to the model (and write the actual warning), so it is not that difficult to deprecate parameters with the setup we already have (especially since we already have the `M_DEPRECATED` error code).

**Comment by jougs** [comment]
@stinebuu @heplesser: it's probably best to discuss the exact policy at the hackathon in a wider group. Maybe we can even come up with a small write-up to circulate among the developers

**Comment by jougs** [comment]
@stinebuu, @heplesser: what's the status of this? Thanks!

**Comment by heplesser** [comment]
@jougs @stinebuu Thanks for the work and discussions! I agree that we can stay with the current framework. We can think about general solutions in context of code generation. I will approve and merge.

**Pull request merged**